plugin/path_config_rotate_root.go

```
@@ -0,0 +1,130 @@
+	// call)
+	creds, err := gcputil.Credentials(cfg.CredentialsRaw)
+	if err != nil {
+		return nil, errwrap.Wrapf("credentials are invalid: {{err}}", err)
+	}
+
+	// Generate a new service account key
+	iamAdmin, err := b.IAMAdminClient(req.Storage)
+	if err != nil {
+		return nil, errwrap.Wrapf("failed to create iam client: {{err}}", err)
+	}
+
+	saName := "projects/-/serviceAccounts/" + creds.ClientEmail
+	newKey, err := iamAdmin.Projects.ServiceAccounts.Keys.
+		Create(saName, &iam.CreateServiceAccountKeyRequest{
+			KeyAlgorithm:   "KEY_ALG_RSA_2048",
+			PrivateKeyType: "TYPE_GOOGLE_CREDENTIALS_FILE",
+		}).
+		Context(ctx).
+		Do()
+	if err != nil {
+		return nil, errwrap.Wrapf("failed to create new key: {{err}}", err)
+	}
+
+	// Base64-decode the private key data (it's the JSON file)
+	newCredsJSON, err := base64.StdEncoding.DecodeString(newKey.PrivateKeyData)
+	if err != nil {
+		return nil, errwrap.Wrapf("failed to decode credentials: {{err}}", err)
+	}
+
+	// Verify creds are valid
+	newCreds, err := gcputil.Credentials(string(newCredsJSON))
+	if err != nil {
+		return nil, errwrap.Wrapf("api returned invalid credentials: {{err}}", err)
+	}
+
+	// Update the configuration
+	cfg.CredentialsRaw = string(newCredsJSON)
+	entry, err := logical.StorageEntryJSON("config", cfg)
+	if err != nil {
+		return nil, errwrap.Wrapf("failed to generate new configuration: {{err}}", err)
+	}
+	if err := req.Storage.Put(ctx, entry); err != nil {
+		return nil, errwrap.Wrapf("failed to save new configuration: {{err}}", err)
+	}
+
+	// Delete the old service account key
+	oldKeyName := fmt.Sprintf("projects/%s/serviceAccounts/%s/keys/%s",
+		creds.ProjectId,
+		creds.ClientEmail,
+		creds.PrivateKeyId)
+	if _, err := iamAdmin.Projects.ServiceAccounts.Keys.
+		Delete(oldKeyName).
+		Context(ctx).
+		Do(); err != nil {
+		return nil, errwrap.Wrapf("failed to delete old service account key: {{err}}", err)
+	}
+
+	// Clear caches to pick up the new credentials
+	b.ClearCaches()
```

I feel this should probably go before the delete, otherwise on a delete fail we'll be in a inconsistent state between cache and store.

---

I moved it to go before delete. I think we should return an error if the delete fails. Users don't look at warnings (usually).

---

👍  Returning an error is fine now that it clearly indicates that the rotation did occur.